Avoid Option<DeviceType> in all structs

mhammond opened the pull request:
Fixes #4860 

It's unfortunate that anyone who deserializes needs to remember to add the custom default path, but we can't add `#[serde(default=...)]` to the enum itself because that's only supported for structs. A default deserialize implementation sounds a little like overkill.

Welcome all feedback, so I'm just marking this as a draft now.

bendk commented on components/sync15/src/client_types.rs:
```diff
@@ -24,60 +24,70 @@ pub struct RemoteClient {
+                "tablet" => DeviceType::Tablet,
+                "vr" => DeviceType::VR,
+                "tv" => DeviceType::TV,
+                _ => DeviceType::Unknown,
```
Should strings that we don't know about map to `DeviceType::Unknown`?  It makes logical sense, but it seems like there could be issues with syncing:

  - Client A sets a field to `DeviceType::BrainImplant`
  - Client B is running an older FF version.  It syncs, doesn't recognize that variant, so it uses `DeviceType::Unknown`
  - Client B syncs again and uploads the record to the server
  - Client A syncs, updates their value to `DeviceType::Unknown`

I'm not sure this can happen right now, but maybe something like that could happen in the future.  Maybe it would be better to throw an error in that case and make client B ignore the record.  I'm sure what's best, so whatever you think is fine with me.  It should be documented in any case though.

skhamis replied:
I'm slightly partial to how it is since that device type technically is "unknown" at that point, and any other way might get a little messy trying to support old formats, etc. However, I do agree documenting it would help us down the line.

bendk reviewed (approved):
I like it this way.

I was trying to think of ways to avoid hand-implementing `Deserialize` and `Serialize`, but then decided that I like the impls.  It makes everything clear and explicit.

skhamis reviewed (approved):
Nice refactor! Def feels cleaner without the option. Hopefully this also makes it easier to consolidate this type to one central place soon!

mhammond commented:
Having unknown be, say, `Unknown(String)` seems ideal, but I'm really not sure how that flies with mobile.

mhammond commented:
I update this in the following way:
* Moved DeviceType to a new device_type.rs and added many comments about the semantics for serializing etc.
* Deserialize "phone" as `DeviceType::Mobile` given https://searchfox.org/mozilla-central/rev/a156a65ced2dae5913ae35a68e9445b8ee7ca457/services/sync/modules/engines/clients.js#292 - I doubt we hit that in practice, but I can't see anything wrong with doing this.

I also tested iOS and Android (although I failed to get Fenix working for reasons unrelated to this. So it seems fine to land.
